### ISSUE TITLE
Accessibility:  Invisible elements "bleed" through stack [Bug] 

### ISSUE THREAD

**Issue by geeetarcode** [xamarin/Xamarin.Forms#12631]
### Description

While implementing ADA compliance on iOS, finding that elements on lower pages of the stack are "bleeding" to the top of the stack. We are working with a 3rd-party Compliance company, who has thoughts about it and how to fix the issue, which will require a Xamarin fix. 

### Steps to Reproduce

1. Have multiple Pages on the stack with multiple elements that are labeled with AutomationProperties.SetIsInAccessibleTree, AutomationProperties.SetName
2. Turn on VoiceOver
3. Notice that invisible elements "bleed through" from Pages under the top of the stack.

### Expected Behavior

Only elements that are assigned with ADA code at the Page at the top of the stack should be responsive to VoiceOver.

### Actual Behavior

Invisible elements from Pages under the top of the stack are triggered and spoken with VoiceOver

### Basic Information

- Version with issue: 4.7.0.1179
- Last known good version: None
- IDE: Visual Studio for Mac
- Platform Target Frameworks: 
  - iOS:  iOS 14
  - Android: 11
- Nuget Packages:
- Affected Devices: iOS

### Workaround

Per the Third-Party ADA Compliance company:

The project seems to have no mention of ‘accessibilityViewIsModal’ for UIViews on iOS (https://developer.apple.com/documentation/objectivec/nsobject/1615089-accessibilityviewismodal). That is one of the two solutions for content bleeding through views. Likely it was never added because I don’t believe there is a direct analog for Android. The other is using UINavigationControllers correctly with UIViewControllers in their stack.
 
As for the accessibiltyViewIsModel it would have to be added in these places:
 
Where the native UIViews’ accessibility vars are being set:
https://github.com/xamarin/Xamarin.Forms/blob/f35ae07a0a8471d255f7a1ebdd51499e10e0a4cb/Xamarin.Forms.Platform.iOS/Extensions/AccessibilityExtensions.cs

And AutomationProperties:
https://github.com/xamarin/Xamarin.Forms/blob/f35ae07a0a8471d255f7a1ebdd51499e10e0a4cb/Xamarin.Forms.Core/AutomationProperties.cs
 
As to what would be likely the best fix is ensuring UIViewControllers are inside of UINavigationControllers so VoiceOver knows they are placed in the stack correctly:
 
UIViewController’s analog seems to be PlatformRenderer or at least that seems like the base class. It seems there are many Renderer classes which makes sense because UIViewController is the base class for many important iOS classes. UINavigationController seems to have the analog of NavigationRenderer. So my question is, on the places where you seem to have bleed-through do you have PlatformRenderers that are on the stack of a NavigationRenderer?

**Comment by rachelkang** [comment]
Hi, @geeetarcode - thanks for submitting this issue. I tried to reproduce this bug but wasn't able to - I put various labels on 3 different pages, and when I'm on the top-most page, VoiceOver only reads the elements on the current page (and none from any underlying pages).

Would you be able to share a sample project that reproduces the issue you're experiencing or edit this one I've tested with here, so that we can better assist you? Thanks :)
[Issue12631.zip](https://github.com/xamarin/Xamarin.Forms/files/5478164/Issue12631.zip)

**Comment by geeetarcode** [comment]
Hi @rachelkang . Unfortunately I can't give an example with the current codebase due to NDA. But, did you by chance try it with a subclass of a Page or a View that was acting as a page? I was just in the codebase and am seeing the same issue with a View on top of a Page. I'd think that whatever is on the topmost view would be the one with ADA focus? Or am I thinking about VoiceOver in Xamarin incorrectly/making bad assumptions?

**Comment by rachelkang** [comment]
@geeetarcode no worries, totally understand! Since I'm not sure exactly how you subclassed your pages or structured your views, would you mind editing the zipped project I shared above ([Issue12631.zip](https://github.com/xamarin/Xamarin.Forms/files/5478430/Issue12631.zip)) to reflect the basic page structure you're thinking of? (no need to share any of the actual code you're working on)  It's possible that the issue is specific to the controls you used on the various pages, or the specific way in which you've created the pages, and I want to make sure we can address your specific case appropriately.

And yes, I agree - the VoiceOver definitely should be reading only the elements on the current / visible page!

**Comment by magoolation** [comment]
As a voiceover user I would like to clarify some points:
A page and a view are two conpletely different things.
A page is a new viewcontroller and receive exclusivity for the focus.
A view is a overlay and unless it is marked as a modal or other elenents are removed from the accessibility tree all views still visible to voiceover

It is not a xamarin specific issue. It is how accessibility abd in this case voiceover works
To solve the problem in your codebase you can apply some native properties using Effects to access the native control

Obter o Outlook para iOS<https://aka.ms/o0ukef>
________________________________
De: Rachel Kang <notifications@github.com>
Enviado: Monday, November 2, 2020 7:48:38 PM
Para: xamarin/Xamarin.Forms <Xamarin.Forms@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Assunto: Re: [xamarin/Xamarin.Forms] Accessibility: Invisible elements "bleed" through stack [Bug] (#12631)


@geeetarcode<https://github.com/geeetarcode> no worries, totally understand! Since I'm not sure exactly how you subclassed your pages or structured your views, would you mind editing the zipped project I shared above (Issue12631.zip<https://github.com/xamarin/Xamarin.Forms/files/5478430/Issue12631.zip>) to reflect the basic page structure you're thinking of? (no need to share any of the actual code you're working on) It's possible that the issue is specific to the controls you used on the various pages, or the specific way in which you've created the pages, and I want to make sure we can address your specific case appropriately.

And yes, I agree - the VoiceOver definitely should be reading only the elements on the current / visible page!

**Comment by geeetarcode** [comment]
Thanks for the feedback. @magoolation , a view being different from a page is understood. I think I found, partially at least, what was going on in the codebase. It seems that we had pop up windows that were coded with a long line of inheritance that seem to be rooted as views rather than pages...thus, there is bleed from the pages below them. It makes sense in some cases to just change these over to be pages, but not in all cases, as it will ruin the non-ADA experience for sighted users. It seems there will have to be a bit of a compromise in the ADA coding to give the best experience possible for _all_ users. If I run into anything different from this issue, I'll be sure to update this thread. 

Thanks for the input, all.

**Comment by mevdev** [comment]
I think the main thing is being able to set accessibilityViewIsModal on a UIView.

That bool should be added so Views can become overlays in Xamarin.Forms.

**Comment by geeetarcode** [comment]
To echo what @mevdev  was saying, it does make more sense to me to have the topmost _view_ have the focus of its elements rather than the topmost _page_, which seems to be the current behavior.

**Comment by uday-lucky** [comment]
@geeetarcode @rachelkang  I faced the same issue in iOS!! Hidden controls are getting accessed when you turn on voice over!!( Which are under the stack). This issue has been fixed in xamarin forms 4.8 . I updated it to 4.8 , now this issue is gone!! 😀 Try to update xf version and give a try

**Comment by rachelkang** [comment]
Hi all! If you haven't already, I recommend updating to the latest version Xamarin.Forms or at least to Xamarin.Forms 5.0.0.2083, which has many accessibility improvements. These improvements include an experimental accessibility flag that addresses a lot of improvements in Android accessibility and corrects behavior like the ones described in this issue.

To leverage this flag on the latest versions of Xamarin Forms, the following code can be inserted in the App.xaml.cs file, following the InitializeComponent call and before the MainPage is set: `Xamarin.Forms.Device.SetFlags(new List<string> { "Accessibility_Experimental"});`

Let us know how that works for you!
